scripts/test-network-stalls.go

```diff
@@ -0,0 +1,248 @@
+// Copyright (C) 2019 Storj Labs, Inc.
+// See LICENSE for copying information.
+
+package main
+
+import (
+	"bytes"
+	"context"
+	"flag"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"regexp"
+	"strconv"
+	"strings"
+	"syscall"
+	"time"
+
+	"github.com/zeebo/errs"
+)
+
+var (
+	numTries      = flag.Int("num-tries", 20, "number of tries to cause a hang")
+	bucketName    = flag.String("bucket", "bukkit", "name of bucket to use for test")
+	fileSize      = flag.Int64("file-size", 5*1024*1024, "size of test file to use")
+	deleteTimeout = flag.Duration("timeout", 60*time.Second, "how long to wait for a delete to succeed or time out")
+
+	tryAgain = errs.New("test needs to run again")
+)
+
+func makeRandomContentsFile(path string, size int64) (err error) {
+	inFile, err := os.Open("/dev/urandom")
+	if err != nil {
+		return err
+	}
+	defer func() {
+		err = errs.Combine(err, inFile.Close())
+	}()
+	outFile, err := os.Create(path)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		err = errs.Combine(err, outFile.Close())
+	}()
+	if _, err := io.CopyN(outFile, inFile, size); err != nil {
+		return err
+	}
+	return nil
+}
+
+// ErrorWithOutput returns an exit error with all collected output.
+type ErrorWithOutput struct {
+	*exec.ExitError
+	Output []byte
+}
+
+func (e *ErrorWithOutput) Error() string {
+	return e.ExitError.Error() + "\nOutput: " + string(e.Output)
+}
+
+type uplinkRunner struct {
+	execName  string
+	configDir string
+	logLevel  string
+}
+
+func (ur *uplinkRunner) run(ctx context.Context, stdout, stderr io.Writer, args ...string) error {
+	cmdArgs := []string{"--config-dir", ur.configDir, "--log.level", ur.logLevel}
+	cmdArgs = append(cmdArgs, args...)
+	cmd := exec.CommandContext(ctx, ur.execName, cmdArgs...)
+	cmd.Stdout = stdout
+	cmd.Stderr = stderr
+	return cmd.Run()
+}
+
+func (ur *uplinkRunner) GetOutputCtx(ctx context.Context, args ...string) ([]byte, error) {
```

Comment:
I think it should be sufficient to have one func called `Run(ctx context.Context, args ...string)`... and ignore the output where you don't need it. Having 4 things that roughly do the same thing is slightly confusing.

Reply:
I had thought this was the norm in Go (see strings.Split..., bytes.Trim..., regexp.Regexp's Find... methods, exec.Command/exec.CommandContext, and most of strconv)

But ok, sure, this isn't a general-purpose library :smile:

Reply:
The reason for `Command` and `CommandContext` is because of backwards compatibility. `context` was added after `exec` package was designed.